burn-dataset/src/transform/sampler.rs

```diff
@@ -17,22 +26,58 @@ where
     D: Dataset<I>,
     I: Send + Sync,
 {
-    /// Creates a new sampler dataset.
+    /// Creates a new sampler dataset with replacement.
     pub fn new(dataset: D, size: usize) -> Self {
         let rng = Mutex::new(StdRng::from_entropy());
 
         Self {
             dataset,
             size,
             rng,
+            state: None,
+            input: PhantomData,
+        }
+    }
+
+    /// Creates a new sampler dataset with replacement.
+    pub fn with_replacement(dataset: D, size: usize) -> Self {
+        Self::new(dataset, size)
+    }
+
+    /// Creates a new sampler dataset without replacement.
+    pub fn without_replacement(dataset: D, size: usize) -> Self {
+        let rng = Mutex::new(StdRng::from_entropy());
+        let state = Mutex::new((0..dataset.len()).collect());
+
+        Self {
+            dataset,
+            size,
+            rng,
+            state: Some(state),
             input: PhantomData,
         }
     }
 
-    /// Generates random index using uniform distribution (0, dataset.len()).
     fn index(&self) -> usize {
+        match &self.state {
+            Some(state) => {
+                let mut state = state.lock().unwrap();
+
+                if state.len() == 0 {
+                    // Refill the state using the same vector.
+                    (0..self.dataset.len()).for_each(|i| state.push(i));
+                }
+
+                let index = self.index_size(state.len());
+                state.remove(index)
```

Comment:
`remove` operation can be costly `O(N)`. See https://doc.rust-lang.org/std/vec/struct.Vec.html#method.remove

I recommend we change the algorithm to shuffle the list of `state` once (using something like https://docs.rs/shuffle/latest/shuffle/)  and use a concurrent data structure to pop a value, perhaps something like: https://github.com/smol-rs/concurrent-queue.  This way you amortize randomization and reduce locking. 

Also, several locks can lead to deadlocks. I didn't work out if it could actually happen in this instance, but it's possible.

Reply:
We should shuffle the array once and pop until there is not item and just refill it. I don't think we need an external crate to do it, but I agree we should change the current implementation to avoid costly operations.